File: src/clients/redis-client.js
const redis = require("redis");
const fs = require('fs');

class RedisClient {
  constructor(numOfInstances = 1, resultFilename = 'results-redis.txt', totalOperations = 10000) {
    this.numOfInstances = numOfInstances;
    this.config = this.generateConfig();
    this.resultFilename = resultFilename;
    this.totalOperations = totalOperations;
  }

  generateConfig() {
    let config = {};

    if (this.numOfInstances === 1) {
      config.url = `redis://127.0.0.1:6370`;
      return config;   
    }
    
    config = {
      rootNodes: []
    };
  
    for (let i = 0; i < this.numOfInstances; i++) {
      config.rootNodes.push({
        url: `redis://127.0.0.1:637${i}`,
      });
    }
  
    return config;
  }

  createInstance(config) {
    if(this.numOfInstances === 1) {
      return redis.createClient(config);
    }

    return redis.createCluster(config);
  }

  benchmarkWrite(dataSize, show=false) {
    return new Promise((resolve) => {
      const data = Buffer.alloc(dataSize, 'a').toString();
      let totalTimeTaken = 0;
      this.keys = [];
      const operations = [];

      for (let i = 0; i < this.totalOperations; i++) {
        const key = `key-${dataSize}-${i}`;
        this.keys.push(key);
        
        operations.push(this.client.set(key, data));
      }

      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
        const end = process.hrtime.bigint();
        totalTimeTaken += Number(end - start) / 1e6;
        const averageTime = totalTimeTaken / this.totalOperations;
        if(show) {
          console.info(`Redis WRITE ASYNC for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Redis, WRITE ASYNC,' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }
        resolve();
      })
    })
  }

  benchmarkRead(dataSize, show=false) {
    return new Promise((resolve) => {
      const keys = this.keys
                    .map(val => ({ val, sort: Math.random() }))
                    .sort((a, b) => a.sort - b.sort)
                    .map(obj => obj.val);

      let operations = [];

      operations = keys.map(key => {
        return this.client.get(key)
      });

      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
        const end = process.hrtime.bigint();

        const averageTime = (Number(end - start) / 1e6) / this.keys.length;
        if(show) {
          console.info(`Redis READ ASYNC for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Redis, READ ASYNC,' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }
        resolve();
      })
    })
  }

  async connect() {
    this.client = this.createInstance(this.config);
    await this.client.connect();
    console.log("Connected to Redis");
    return this.client;
  }

  async disconnect() {
    await this.cleanup();
    await this.client.quit();
    console.log("Redis Client Disconnected");
  }

  async cleanup() {
    for (let key of this.keys) {
      await this.client.del(key);
    }
  }
}

module.exports = RedisClient;

File: python_clients/redis_client.py
from redis import Redis as redisClient
from redis.cluster import RedisCluster as redisCluster
from redis.cluster import ClusterNode
import random
import time

class RedisClient:
  def __init__(self, numOfInstances = 1, resultFilename = 'res-redis-py.txt', totalOperations = 10000, totalProcesses = 1):
    self.numOfInstances = numOfInstances
    self.config = self.generateConfig(self.numOfInstances)
    self.resultFilename = resultFilename
    self.totalOperations = totalOperations
    self.totalProcesses = totalProcesses
    self.keys = []
    self.client = None


  def generateConfig(self, numOfInstances):
    config = []

    # config = [
    #   ClusterNode('172.23.0.7', 6379),
    #   ClusterNode('172.23.0.8', 6379),
    #   ClusterNode('172.23.0.3', 6379),
    #   # ClusterNode('172.22.0.14', 6379),
    #   # ClusterNode('172.22.0.30', 6379),
    #   # ClusterNode('172.22.0.7', 6379),
    #   # ClusterNode('172.22.0.4', 6379),
    #   # ClusterNode('172.22.0.21', 6379),
    #   # ClusterNode('172.22.0.5', 6379),
    #   # ClusterNode('172.22.0.12', 6379),
    # ]
    for i in range(numOfInstances):
      config.append(ClusterNode('localhost', 6370 + i))

    
    return config


  def connect(self):

    if len(self.config) == 1:
      self.client = redisClient(host = 'localhost', port = 6370)
    else:
      self.client = redisCluster(startup_nodes = self.config)

    # print("Connected to Redis")

  def connect(self, instanceNumber):
    self.client = redisClient(host = 'localhost', port = 6370 + instanceNumber)

  def disconnect(self):
    self.client.close()
    # print("Disconnected from Redis")

  def benchmarkWrite(self, dataSize, show=False):
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Redis WRITE SINGLE {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Redis WRITE SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkWriteMultiprocess(self, dataSize, show=False):
    self.connect()
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Redis WRITE MULTI {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Redis WRITE MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()


  def benchmarkWriteMultiprocessOld(self, dataSize, offset, show=False, processNum=1):
    self.connect()
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(offset, offset + self.totalOperations//self.totalProcesses):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.client.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Redis WRITE MULTI {dataSize} {averageTime:.12f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Redis WRITE MULTI {dataSize} {averageTime:.12f}\n')
    resultFile.close()

    self.disconnect()


  def benchmarkRead(self, dataSize, show=False):
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.client.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Redis READ SINGLE {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Redis READ SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()

  
  def benchmarkReadMultiprocess(self, dataSize, show=False, processNum=1):
    self.connect()
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.client.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Redis READ MULTI {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Redis READ MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()


  def cleanup(self):
    self.client.flushall()